### ISSUE TITLE
Support running driver tests on non-linux OSes

### ISSUE THREAD

**Issue by bzz** [bblfsh/sdk#370]
Right now the integration test part of the `bblfsh-sdk test` flow relies on ability to share a `/var/run/bblfshctl.sock` between Host and an a container running `bblfshd`, so that a local built driver image could be installed in `bblfhsd` from that docker instance (instead of usual dockerHub) [using the `docker-daemon` transport](https://doc.bblf.sh/using-babelfish/advanced-usage.html#adding-drivers-from-the-local-docker-daemon).

That is a workaround that works *only on linux* (as other OSes simiply do not use sockets and communicate though TCP to actual docker daemon running in linux VM), found during previous attempts to archive similar thing under https://github.com/bblfsh/bblfshd/issues/74.

What is requited here, is to optionally _change/override the way how `bblfshd` talks to docker_ in `bblfhsh-sdk` case to the communication over TCP (similar to how it is possible to override how `bblfshctl` talks to `bblfhsd` from socket to TCP using `ctl-network`/`ctl-address` [CLI options](https://github.com/bblfsh/bblfshd/blob/83166ea0087bfe20c24fc471309f70f422383198/cmd/bblfshctl/cmd/base.go#L16) )

As one can see from `vendor/github.com/containers/image/docker/daemon/daemon_src.go:38` - the old version of dependency we are using inside `runtime` does not support it, but it seems to be possible with [the latest one](https://github.com/containers/image/blob/5144ced37a1b21b63c6ef605e56811e29a687528/docker/daemon/client.go#L18).

For convenience, it would make sense to expose this options as both ENV vars and CLI args with the same names, so they could be easily customized on bblshd startup and applied in [integration test of `bblfsh-sdk test`](https://github.com/bblfsh/sdk/blob/5a1a268859490ec2772d8106a331be1f7d0d85a3/build/dev.go#L91).

**Comment by creachadair** [comment]
I would advocate that instead of making this switchable, we should switch it on _all_ platforms to use a loopback TCP socket. As far as I can tell that is supported on all OSs we are likely to encounter, and I think there is almost no advantage to having multiple options here.

Of course, we might have to support both during a transition, but I think the goal state should be to wind up with only one method of connection. Loopback is nice because it's still in-kernel, and as long as we are careful not to bind only to the loopback address does not expose the server more widely than the socket does.

**Comment by dennwc** [comment]
I think it's better to rely on what Docker client provides - it will automatically select the connection based on OS and environment.

Forcing it to loopback will most probably force everyone to reconfigure Docker on Linux.

**Comment by creachadair** [comment]
> I think it's better to rely on what Docker client provides - it will automatically select the connection based on OS and environment.
> 
> Forcing it to loopback will most probably force everyone to reconfigure Docker on Linux.

Agreed. Sorry, I wrote imprecisely: I don't mean that we should hard-code loopback in the code itself, just that we should make it use TCP and set it up to use loopback in our own usage.

**Comment by dennwc** [comment]
Sorry, I don't see how it's different.

If developer installs Docker on Linux, it will use Unix socket by default, and won't listen on any TCP port. Then, after installing an SDK, all commands will try to use TCP and will fail.

**Comment by creachadair** [comment]
> Sorry, I don't see how it's different.
> 
> If developer installs Docker on Linux, it will use Unix socket by default, and won't listen on any TCP port. Then, after installing an SDK, all commands will try to use TCP and will fail.

Maybe we're talking about different things? I thought this issue was about the communication between the host and the Babelfish daemon, not the docker server. I don't think it matters what Docker itself uses to communicate with its CLI, but it does matter (if I read this correctly) how the host client communicates with the daemon.

Did I miss something important?

**Comment by dennwc** [comment]
Sorry for the confusion, the issue is about `bblfshd` communication with the Docker daemon on the host.

As a part of the integration test suite, we run a `bblfshd` daemon in Docker. We build a dev version of the driver locally (in Docker) and install it into the running `bblfshd` instance. For this to work, `bblfshd` needs to know how to communicate with the Docker daemon to pull the container image from it.

For Linux, Docker runs on the host and is accessible via the Unix socket, which is exposed as a volume to the `bblfshd` container.

But for macOS, Docker runs in the VM and cannot be accessed the same way as on Linux.

**Comment by creachadair** [comment]
Ah, okay, I misunderstood the layering here. In that case, you can probably ignore most of what I said above.  Thanks for the clarification!

**Comment by bzz** [comment]
Stranger then fiction: while integration tests hang on 10 min timeout with logs

```
$ time CI="yes" go run vendor/gopkg.in/bblfsh/sdk.v2/cmd/bblfsh-sdk/main.go test
 ....

docker run --rm --privileged -v /var/run/docker.sock:/var/run/docker.sock bblfsh/bblfshd
+ docker exec sha256:36d10eb1a380839c848130dd116f77d95f480bf78c59959490f0e5cb566053fa bblfshctl driver install javascript docker-daemon:sha256:36d10eb1a380839c848130dd116f77d95f480bf78c59959490f0e5cb566053fa
```

If I manually list installed drivers in that container - it seems to have a driver installed 😕 

```
docker exec 61158be3fa9b bblfshctl driver list                                                                                          
+------------+---------------------------------------------------------------------------------------+--------------+--------+-----------+--------+-------------+---------------+
|  LANGUAGE  |                                         IMAGE                                         |   VERSION    | STATUS |  CREATED  |   OS   |     GO      |    NATIVE     |
+------------+---------------------------------------------------------------------------------------+--------------+--------+-----------+--------+-------------+---------------+
| javascript | docker-daemon:sha256:36d10eb1a380839c848130dd116f77d95f480bf78c59959490f0e5cb566053fa | dev-8c08ccf1 | beta   | 5 minutes | alpine | 1.10-alpine | node:8-alpine |
+------------+---------------------------------------------------------------------------------------+--------------+--------+-----------+--------+-------------+---------------+
Response time 441.6µs
```

**Comment by dennwc** [comment]
Yeah, it seems like a different container. The test runs and kills it's own `bblfshd` instance.

**Comment by creachadair** [comment]
On macOS, it turns out that Docker symlinks `/var/run/docker.sock` to the "real" location under the invoking user's `$HOME`. I manually updated the command to expand the link on the LHS of the bind mapping, and it appears to work:

```shell
docker run --rm --privileged -v $HOME/Library/Containers/com.docker.docker/Data/docker.sock:/var/run/docker.sock bblfsh/bblfshd
```

I hand-patched the SDK to do this evaluation when starting up the driver, and it is able to connect. However, I didn't fully push this change down the stack, and the `bblfsh/bblfshd` image I was using was pre-baked with the old value and failed partway through.

Nonetheless, I think this strategy has the potential to succeed. The only real change I made was to add

```go
// LocalSocket returns the fully expanded path of the Docker socket on the                                                                                                                                      
// current machine, by expanding symbolic links in Socket.                                                                                                                                                      
func LocalSocket() string {
   if v, err := filepath.EvalSymlinks(Socket); err == nil {
      return v
   }
   return Socket // fallback                                                                                                                                                                                    
}
```

and to replace the constant strings with calls to that function.

**Comment by creachadair** [comment]
As you might expect, it turns out to be a little more complicated than that: The code that talks to the daemon is buried inside https://github.com/containers/image/blob/master/docker/daemon/client.go and to control it we would have to plumb through a [`SystemContext`](https://godoc.org/github.com/containers/image/types#SystemContext) carrying the bound `DockerDaemonHost` we want to use. Right now it is set to `nil` which falls back on [`DefaultDockerHost`](https://godoc.org/github.com/docker/docker/client#pkg-constants).

So this is reachable, but will require some thought about how to wire it up. I'll keep at it.

**Comment by creachadair** [comment]
Interestingly, [the `docker` package](https://godoc.org/github.com/docker/docker/client#pkg-constants) suggests maybe we could override this in the environment.

**Comment by creachadair** [comment]
…except, we don't use that package, we use https://godoc.org/github.com/ory/dockertest/docker instead. And the config there doesn't expose the host in the same way (I think).
